### PR TITLE
Enhance the diagnostic column

### PR DESCRIPTION
**Pull request by scottming**
### before

image

image

### After

image

image

The cursor color and highlighting(no highlighting spaces) look better, and we can reach the exact place we want to modify with fewer taps w when using something like: `Lspsaga diagnostic_jump_next`.

### CODE REVIEW

File: apps/common/lib/lexical/math.ex [hunk 1]
```diff
@@ -2,4 +2,11 @@ defmodule Lexical.Math do
   def clamp(number, min_number, max_number) do
     number |> max(min_number) |> min(max_number)
   end
+
+  def count_leading_spaces(str), do: count_leading_spaces(str, 0)
+
+  def count_leading_spaces(<<c::utf8, rest::binary>>, count) when c in [?\s, ?\t],
```
Comment (scohen) [inline apps/common/lib/lexical/math.ex:8]: I don't think you need the utf8 specificer here, but it's not hurting things.

File: apps/common/lib/lexical/math.ex [hunk 2]
```diff
@@ -2,4 +2,11 @@ defmodule Lexical.Math do
   def clamp(number, min_number, max_number) do
     number |> max(min_number) |> min(max_number)
   end
+
+  def count_leading_spaces(str), do: count_leading_spaces(str, 0)
```
Comment (scohen) [inline apps/common/lib/lexical/math.ex:6]: This function doesn't seem like it belongs in the `Math` module, as it's more of a String thing.

File: apps/common/test/lexical/math_test.exs [hunk 3]
```diff
@@ -28,4 +28,22 @@ defmodule Lexical.MathTest do
...
+      assert Math.count_leading_spaces("hello") == 0
+    end
+
+    test "returns 2 when string has two leading spaces" do
+      assert Math.count_leading_spaces("  hello") == 2
+    end
+
+    test "return 1 when string has one leading tab" do
+      assert Math.count_leading_spaces("\thello") == 1
+    end
```
Comment (scohen) [inline apps/common/test/lexical/math_test.exs:47]: i feel like property tests would be appropriate here.

File: apps/server/lib/lexical/server/project/diagnostics.ex [hunk 4]
```diff
@@ -137,11 +137,23 @@ defmodule Lexical.Server.Project.Diagnostics do
...
+      end
+    end
+
+    defp to_lsp_range(line, column, source_file) do
       elixir_range =
         ExRange.new(
-          ExPosition.new(line_number, column),
-          ExPosition.new(line_number + 1, 1)
+          ExPosition.new(line, column),
+          ExPosition.new(line + 1, 1)
```
Comment (scohen) [inline apps/server/lib/lexical/server/project/diagnostics.ex:156]: line_number is more descriptive; In this codebase, the word `line` refers to a specific data structure.

File: apps/server/lib/lexical/server/project/diagnostics.ex [hunk 5]
```diff
@@ -150,18 +162,11 @@ defmodule Lexical.Server.Project.Diagnostics do
 
         _error ->
           Range.new(
-            start: Position.new(line: line_number, character: 0),
-            end: Position.new(line: line_number + 1, character: 0)
+            start: Position.new(line: line, character: 0),
+            end: Position.new(line: line + 1, character: 0)
```
Comment (scohen) [inline apps/server/lib/lexical/server/project/diagnostics.ex:166]: same comment; change back to `line_number`

File: apps/common/lib/lexical/math.ex (hunk 2 above)
Comment (scottming) [reply apps/common/lib/lexical/math.ex:6]: In your opinion, which file in the 'common' folder do you think is more suitable for placing this `count_leading_spaces` function?

File: apps/common/test/lexical/math_test.exs (hunk 3 above)
Comment (scottming) [reply apps/common/test/lexical/math_test.exs]: Added one via e767c11b7c6fb7798e8c9788830b3c741aef4b38

File: apps/common/lib/lexical/math.ex (hunk 2 above)
Comment (scottming) [reply apps/common/lib/lexical/math.ex]: Ok, I moved to the `Lexical.Text`

File: apps/server/lib/lexical/server/project/diagnostics.ex [hunk 6]
```diff
@@ -137,7 +138,19 @@ defmodule Lexical.Server.Project.Diagnostics do
...
+    defp position_to_range(source_file, line_number) when is_integer(line_number) do
+      line_number = Math.clamp(line_number, 1, SourceFile.size(source_file))
+
+      with {:ok, line_text} <- SourceFile.fetch_text_at(source_file, line_number) do
+        column = Text.count_leading_spaces(line_text) + 1
+        to_lsp_range(line_number, column, source_file)
+      end
+    end
 
+    defp to_lsp_range(line_number, column, source_file) do
```
Comment (scohen) [inline apps/server/lib/lexical/server/project/diagnostics.ex:153]: what happens if we don't make this an LSP range, and instead use a Lexical range?

**Review by scohen — APPROVED** [review]
try my suggestion and then merge.

File: apps/server/lib/lexical/server/project/diagnostics.ex (hunk 6 above)
Comment (scottming) [reply apps/server/lib/lexical/server/project/diagnostics.ex:153]: #103

**Pull request merged**